### PR TITLE
feat(ingestion): Business Glossary# Add domain support in  GlossaryTerm ingestion

### PR DESCRIPTION
**Pull request by mohdsiddique**
Business Glossary

### CODE REVIEW

File: metadata-ingestion/docs/sources/business-glossary/datahub-business-glossary.md [hunk 1]
```diff
@@ -36,6 +36,7 @@ The business glossary source file should be a `.yml` file with the following top
 - **inherits**: (optional) List of **GlossaryTerm** that this term inherits from
 - **contains**: (optional) List of **GlossaryTerm** that this term contains
 - **custom_properties**: A map of key/value pairs of arbitrary custom properties
+- **domains**: (optional) List of domain name or domain urn
```
Comment (gabe-lyons) [inline metadata-ingestion/docs/sources/business-glossary/datahub-business-glossary.md:39]: let's just make it singular for now to match the UI: just one domain at most

File: metadata-ingestion/examples/bootstrap_data/business_glossary.yml [hunk 2]
```diff
@@ -23,6 +23,8 @@ nodes:
         description: Highly Confidential Data
         custom_properties:
           is_confidential: true
+        domains:
```
Comment (gabe-lyons) [inline metadata-ingestion/examples/bootstrap_data/business_glossary.yml:26]: rename to domain

File: metadata-ingestion/tests/integration/business-glossary/business_glossary.yml [hunk 3]
```diff
@@ -26,6 +26,9 @@ nodes:
         description: Highly Confidential Data
         custom_properties:
           is_confidential: true
+        domains:
+          - en.wikipedia.org
```
Comment (gabe-lyons) [inline metadata-ingestion/tests/integration/business-glossary/business_glossary.yml:30]: this is not same concept as web domain- a better example would be `Marketing`

File: metadata-ingestion/src/datahub/ingestion/source/metadata/business_glossary.py [hunk 4]
```diff
@@ -284,16 +300,30 @@ def get_mces_from_node(
...
                 defaults=defaults,
                 ingestion_config=ingestion_config,
+                ctx=ctx,
             )
 
 
+def get_domain_class(
+    graph: Optional[DataHubGraph], domains: List[str]
+) -> models.DomainsClass:
+    domain_registry: DomainRegistry = DomainRegistry(
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/metadata/business_glossary.py:310]: domain registry should be an instance variable

File: metadata-ingestion/tests/integration/business-glossary/glossary_events_golden.json [hunk 5]
```diff
@@ -145,6 +145,20 @@
         "runId": "datahub-business-glossary-2020_04_14-07_00_00"
     }
 },
+{
+    "entityType": "glossaryTerm",
+    "entityUrn": "urn:li:glossaryTerm:Classification.Highly Confidential",
+    "changeType": "UPSERT",
+    "aspectName": "domains",
+    "aspect": {
+        "value": "{\"domains\": [\"urn:li:domain:en.wikipedia.org\"]}",
```
Comment (gabe-lyons) [inline metadata-ingestion/tests/integration/business-glossary/glossary_events_golden.json:154]: better example would be `urn:li:domain:Marketing`

File: metadata-ingestion/examples/bootstrap_data/business_glossary.yml [hunk 6]
```diff
@@ -23,6 +23,8 @@ nodes:
...
+          - en.wikipedia.org
```
Comment (hsheth2) [inline metadata-ingestion/examples/bootstrap_data/business_glossary.yml:27]: let's do a better example - maybe "Finance" or "Marketing"

**Review by hsheth2 — CHANGES_REQUESTED** [review]

File: metadata-ingestion/docs/sources/business-glossary/datahub-business-glossary.md (hunk 1 above)
Comment (mohdsiddique) [reply metadata-ingestion/docs/sources/business-glossary/datahub-business-glossary.md]: Done

File: metadata-ingestion/examples/bootstrap_data/business_glossary.yml (hunk 2 above)
Comment (mohdsiddique) [reply metadata-ingestion/examples/bootstrap_data/business_glossary.yml]: done

File: metadata-ingestion/tests/integration/business-glossary/business_glossary.yml (hunk 3 above)
Comment (mohdsiddique) [reply metadata-ingestion/tests/integration/business-glossary/business_glossary.yml]: done

File: metadata-ingestion/tests/integration/business-glossary/glossary_events_golden.json (hunk 5 above)
Comment (mohdsiddique) [reply metadata-ingestion/tests/integration/business-glossary/glossary_events_golden.json]: done

File: metadata-ingestion/examples/bootstrap_data/business_glossary.yml (hunk 6 above)
Comment (mohdsiddique) [reply metadata-ingestion/examples/bootstrap_data/business_glossary.yml]: done

File: metadata-ingestion/src/datahub/ingestion/source/metadata/business_glossary.py (hunk 4 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/metadata/business_glossary.py:310]: Initially, I made it the instance variable, however, it doesn't resolve the domain name to the urn in _domain_registry.get_domain_urn(domain)_ function call. It resolves the domain name to the urn in the constructor call.

To make it an instance variable either I need to change the behavior of DomainRegistry or I need to traverse the business glossary in advance to collect all domains and pass it to the cache argument at the time of construction.

I think changing the behavior of DomainRegistry would be good. Let me know your thoughts both of the solutions will take some time extra to refactor the existing code

Comment (hsheth2) [reply metadata-ingestion/src/datahub/ingestion/source/metadata/business_glossary.py:314]: In that case, let's just leave it as-is - i've added a todo to the code

**Review by hsheth2 — APPROVED** [review]

**Review by jjoyce0510 — APPROVED** [review]
LGTM

**Pull request merged**